samza-core/src/main/scala/org/apache/samza/storage/ContainerStorageManagerUtil.java

```diff
@@ -411,4 +438,292 @@ public static Set<String> getSideInputStoreNames(
+        // 4. Mark new Snapshots to never expire
+        Map<String, CompletionStage<Void>> removeNewSnapshotsTTLFutures =
+            removeTTLNewSnapshots(blobStoreUtil, scms, taskName, jobConfig);
+        // Combined future of delete old snapshots and removeTTL of new snapshots
+        CompletableFuture<Void> removeTTLFuture = CompletableFuture.allOf(FutureUtil.mapToFuture(deleteOldSnapshotsFutures),
+            FutureUtil.mapToFuture(removeNewSnapshotsTTLFutures));
+        // 5. Update taskCheckpoints map with new checkpoint
+        CompletableFuture<Void> checkpointFuture = removeTTLFuture.thenRun(() -> {
+          Checkpoint newCheckpoint = writeNewCheckpoint(taskName, checkpointId, scms, checkpointManager);
+          taskCheckpoints.put(taskName, newCheckpoint);
+        });
+        return checkpointFuture;
+      });
+      return future;
+    }).exceptionally(ex -> {
+      String msg = String.format("Restore for task %s failed with get deleted bob", taskName);
+      throw new SamzaException(msg, ex);
+    });
+  }
+
+  private static CompletableFuture<Map<String, String>> createNewCheckpointAndBackupStores(JobContext jobContext,
+      ContainerModel containerModel, Config config, TaskName taskName, Set<String> storesToBackup,
+      CheckpointId newCheckpointId, File loggedStoreBaseDirectory, BlobStoreManager blobStoreManager,
+      MetricsRegistry metricsRegistry, ExecutorService executor) {
+
+    BlobStoreBackupManagerMetrics blobStoreBackupManagerMetrics = new BlobStoreBackupManagerMetrics(metricsRegistry);
+    BlobStoreBackupManager blobStoreBackupManager =
+        new BlobStoreBackupManager(jobContext.getJobModel(), containerModel, containerModel.getTasks().get(taskName),
+            executor, blobStoreBackupManagerMetrics, config, SystemClock.instance(), loggedStoreBaseDirectory,
+            new StorageManagerUtil(), blobStoreManager);
+
+    // create checkpoint dir as a copy of store dir
+    createCheckpointDirFromStoreDirCopy(taskName, containerModel.getTasks().get(taskName),
+        loggedStoreBaseDirectory, storesToBackup, newCheckpointId);
+    // upload to blob store and return future
+    return blobStoreBackupManager.upload(newCheckpointId, new HashMap<>());
+  }
+
+  private static Checkpoint writeNewCheckpoint(TaskName taskName, CheckpointId checkpointId,
+      Map<String, String> newSCMs, CheckpointManager checkpointManager) {
+    CheckpointV2 oldCheckpoint = (CheckpointV2) checkpointManager.readLastCheckpoint(taskName);
+    Map<SystemStreamPartition, String> checkpointOffsets = oldCheckpoint.getOffsets();
+
+    Map<String, Map<String, String>> oldStateCheckpointMarkers = oldCheckpoint.getStateCheckpointMarkers();
+
+    Map<String, Map<String, String>> newStateCheckpointMarkers = ImmutableMap.<String, Map<String, String>>builder()
+        .put(KafkaChangelogStateBackendFactory.class.getName(), oldStateCheckpointMarkers.get(KafkaChangelogStateBackendFactory.class.getName()))
```

Comment:
What if value isn't present (changelog is disabled)? Should this be adding null checkpoints?

Reply:
Updated a check here. Thanks